app/src/main/java/org/mozilla/focus/fragment/HomeFragment.kt

```diff
@@ -24,13 +25,19 @@ import org.mozilla.focus.telemetry.TelemetryWrapper
 import org.mozilla.focus.telemetry.UrlTextInputLocation
 import org.mozilla.focus.utils.OnUrlEnteredListener
 import android.graphics.BitmapFactory
+import java.lang.ref.WeakReference
 
 private const val COL_COUNT = 5
 
+interface HomeDelegate {
+    fun didPressSettings()
+}
+
 /** The home fragment which displays the navigation tiles of the app. */
 class HomeFragment : Fragment() {
     lateinit var urlBar: LinearLayout
     var onUrlEnteredListener = object : OnUrlEnteredListener {} // default impl does nothing.
+    var delegate: WeakReference<HomeDelegate>? = null
```

Comment:
This weak reference is unnecessary: Java is not reference-counted. Objects are GC'd when they're no longer connected to the main object tree by a strong reference, no matter how strong references they have pointing at them (i.e. so MainActivity <-strong-> HomeFragment can both be GC'd if no one in the main object tree has a reference to either of them).

The case for weak references in Android relates to the Context object: an Activity is a context and it's really big so you don't want to keep around any Contexts that don't need to be there or else OOM looms near. When you do certain actions, like rotating the device, the Activity is recreated. Because it's so easy to recreate the Activity, if you keep strong references to the Activity/Context in objects that outlive the Activity/Context (e.g. an async task), you can end up with multiple Contexts (the new one and the invalid ones in async tasks) taking up a lot of space. It's well documented if you want to do more reading: https://android-developers.googleblog.com/2009/01/avoiding-memory-leaks.html

In this particular case, the Activity lifecycle is a superset of the Fragment lifecycle so we don't need to do weak references.

---

nit: In any case, creating a weak reference is really the responsibility of the class receiving the object, so I'd actually change this var to a `fun setDelegate(arg: ...) { delegate = new WeakReference<...>(arg) }`, even if it's more verbose.

---

You mentioned on Vidyo about trying to make this a function type – it's probably hard to use function types in Java so this seems fine. If you really wanted to try:
- I think you could actually find the function types Kotlin is compiling down to ([like this?](https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.reflect/-k-function/index.html)) and instantiate one of those, but it's probably really verbose and not worth the effort.
- You could create a kotlin file that MainActivity calls into which will set the listener. Maybe an extension function?